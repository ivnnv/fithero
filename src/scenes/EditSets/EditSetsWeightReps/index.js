/* @flow */

export { default } from './EditSetsWeightReps';
