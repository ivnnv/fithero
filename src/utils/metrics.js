/* @flow */

import * as RNLocalize from 'react-native-localize';

import type { WorkoutExerciseWeightRepsType } from '../database/types';
import type { DefaultUnitSystemType } from '../redux/modules/settings';

const oneKgToLb = 2.204622622;

export const toLb = (value: number) => value * oneKgToLb;
export const toKg = (value: number) => value / oneKgToLb;
export const toTwoDecimals = (value: number) => Math.round(value * 100) / 100;

export const getWeight = (
  value: number,
  exercise: ?WorkoutExerciseWeightRepsType,
  defaultUnitSystem: DefaultUnitSystemType
) => {
  const unit =
    exercise &&
    ((exercise.isValid && exercise.isValid()) || !exercise.isValid) &&
    exercise.weight_unit
      ? exercise.weight_unit
      : defaultUnitSystem;
  if (unit !== 'metric') {
    return toLb(value);
  }
  return value;
};

export const getWeightUnit = (
  exercise: ?WorkoutExerciseWeightRepsType,
  defaultUnitSystem: DefaultUnitSystemType
): DefaultUnitSystemType =>
  exercise &&
  ((exercise.isValid && exercise.isValid()) || !exercise.isValid) &&
  exercise.weight_unit
    ? exercise.weight_unit
    : defaultUnitSystem;

export const getDefaultUnitSystemByCountry = (): DefaultUnitSystemType => {
  const country = RNLocalize.getCountry();
  if (
    country === 'US' ||
    country === 'UK' ||
    country === 'MM' ||
    country === 'LR'
  ) {
    return 'imperial';
  }
  return 'metric';
};
