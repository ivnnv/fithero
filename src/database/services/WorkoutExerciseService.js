/* @flow */

import realm from '../index';
import type {
  AddWorkoutExerciseWeightRepsType,
  WorkoutExerciseSchemaType,
} from '../types';
import { dateToWorkoutId, getTimeAgo } from '../../utils/date';
import { WORKOUT_EXERCISE_SCHEMA_NAME } from '../schemas/WorkoutExerciseSchema';
import { WORKOUT_SCHEMA_NAME } from '../schemas/WorkoutSchema';

export const getWorkoutExerciseById = (id: string) =>
  realm.objects(WORKOUT_EXERCISE_SCHEMA_NAME).filtered(`id = $0`, id);

export const addExercise = (exercise: AddWorkoutExerciseWeightRepsType) => {
  realm.write(() => {
    const workoutId = dateToWorkoutId(exercise.date);
    let workout = realm.objectForPrimaryKey(WORKOUT_SCHEMA_NAME, workoutId);

    if (!workout) {
      workout = realm.create(WORKOUT_SCHEMA_NAME, {
        id: workoutId,
        date: exercise.date,
      });
    }

    workout.exercises.push({
      ...exercise,
      sort: workout ? workout.exercises.length + 1 : 0,
    });
  });
};

export const deleteWorkoutExercise = (exercise: WorkoutExerciseSchemaType) => {
  const workoutId = dateToWorkoutId(exercise.date);
  realm.delete(exercise);
  // Now we check if workout needs to be deleted too
  const workout = realm.objectForPrimaryKey(WORKOUT_SCHEMA_NAME, workoutId);
  if (workout.exercises.length > 0) {
    // If workout was not deleted, but one exercise yes, let's fix the sort
    workout.exercises.forEach((e, i) => {
      e.sort = i + 1;
    });
  } else if (!workout.comments) {
    realm.delete(workout);
  }
};

export const getExercisesByType = (type: string) =>
  realm
    .objects(WORKOUT_EXERCISE_SCHEMA_NAME)
    .filtered('type = $0', type)
    .sorted([['date', true]]);

export const getRecentExercises = (date: string) => {
  const monthAgo = getTimeAgo(date, 'month', 1);
  return realm
    .objects(WORKOUT_EXERCISE_SCHEMA_NAME)
    .filtered('date >= $0 AND TRUEPREDICATE DISTINCT(type)', monthAgo);
};
