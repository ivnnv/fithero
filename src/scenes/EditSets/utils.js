/* @flow */

import { Keyboard } from 'react-native';

import {
  addSet,
  getLastSetByType,
  updateSet,
} from '../../database/services/WorkoutSetService';
import type {
  ExerciseCategoryType,
  WorkoutExerciseWeightRepsType,
  WorkoutSetWeightRepsType,
} from '../../database/types';
import { getWeight, toTwoDecimals } from '../../utils/metrics';
import type { DefaultUnitSystemType } from '../../redux/modules/settings';
import {
  extractSetIndexFromDatabase,
  getExerciseSchemaId,
  getSetSchemaId,
} from '../../database/utils';
import { toDate } from '../../utils/date';
import { addExercise } from '../../database/services/WorkoutExerciseService';

export const getLastSet = (
  exercise: ?WorkoutExerciseWeightRepsType,
  exerciseKey: string
) => {
  if (exercise) {
    return exercise.sets[exercise.sets.length - 1];
  } else {
    const sets = getLastSetByType(exerciseKey || exercise?.type);
    if (sets.length > 0) {
      return sets[0];
    }

    return null;
  }
};

export const getLastWeight = (
  exercise: ?WorkoutExerciseWeightRepsType,
  lastSet: ?WorkoutSetWeightRepsType,
  unit: DefaultUnitSystemType
) => {
  const defaultWeight = unit === 'metric' ? 20 : 45;

  const lastWeight = lastSet
    ? toTwoDecimals(getWeight(lastSet.weight, exercise, unit))
    : defaultWeight;

  return lastWeight.toString();
};

type AddSetFromInputType = {
  day: string,
  exerciseKey: string,
  exercise: ?WorkoutExerciseWeightRepsType,
  category: ExerciseCategoryType,
  partialSet: {
    weight: number,
    reps: number,
  },
  weight_unit: ?DefaultUnitSystemType,
  updateId?: string,
};

export const addSetFromInput = ({
  day,
  exerciseKey,
  exercise,
  category,
  partialSet,
  weight_unit = null,
  updateId,
}: AddSetFromInputType) => {
  Keyboard.dismiss();

  const date = toDate(day);
  const newSet = {
    date,
    type: exerciseKey,
    category,
    ...partialSet,
  };

  if (!exercise) {
    const exerciseIdDb = getExerciseSchemaId(day, exerciseKey);
    const date = toDate(day);
    const newExercise = {
      id: exerciseIdDb,
      sets: [
        {
          ...newSet,
          id: getSetSchemaId(day, exerciseKey, 1),
        },
      ],
      date,
      type: exerciseKey,
      weight_unit,
      category,
    };

    addExercise(newExercise);
  } else if (!updateId) {
    const lastId = exercise.sets[exercise.sets.length - 1].id;
    const lastIndex = extractSetIndexFromDatabase(lastId);

    addSet({
      ...newSet,
      id: getSetSchemaId(day, exerciseKey, lastIndex + 1),
    });
  } else if (updateId) {
    updateSet({
      ...newSet,
      id: updateId,
    });
  }
};
