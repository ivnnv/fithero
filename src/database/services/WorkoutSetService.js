/* @flow */

import type {
  ExerciseCategoryType,
  WorkoutSetSchemaType,
  WorkoutSetWeightRepsType,
} from '../types';
import realm from '../index';
import { getExerciseSchemaIdFromSet } from '../utils';
import { deleteWorkoutExercise } from './WorkoutExerciseService';
import { getFirstAndLastWeekday, getToday } from '../../utils/date';
import { WORKOUT_SET_SCHEMA_NAME } from '../schemas/WorkoutSetSchema';
import { WORKOUT_EXERCISE_SCHEMA_NAME } from '../schemas/WorkoutExerciseSchema';

export const getMaxWeightByType = (type: string) =>
  realm
    .objects(WORKOUT_SET_SCHEMA_NAME)
    .filtered('type = $0', type)
    .sorted([['weight', true], ['reps', true], 'date', 'id']);

export const getMaxRepByType = (type: string) =>
  realm
    .objects(WORKOUT_SET_SCHEMA_NAME)
    .filtered('type = $0', type)
    .sorted([['reps', true], ['weight', true], 'date', 'id']);

export const getMaxTimeByType = (type: string) =>
  realm
    .objects(WORKOUT_SET_SCHEMA_NAME)
    .filtered('type = $0', type)
    .sorted([['time', true], 'date', 'id']);

export const addSet = (set: WorkoutSetSchemaType) => {
  realm.write(() => {
    const exerciseId = getExerciseSchemaIdFromSet(set.id);
    const exercise = realm.objectForPrimaryKey(
      WORKOUT_EXERCISE_SCHEMA_NAME,
      exerciseId
    );
    exercise.sets.push(set);
  });
};

export const updateSet = (updatedSet: WorkoutSetWeightRepsType) => {
  realm.write(() => {
    const { id, ...restUpdateSet } = updatedSet;
    const set = realm.objectForPrimaryKey(WORKOUT_SET_SCHEMA_NAME, id);
    Object.keys(restUpdateSet).forEach(k => {
      set[k] = restUpdateSet[k];
    });
  });
};

export const deleteSet = (setId: string) => {
  // Database, if last set, delete exercise, if last exercise, delete workout
  realm.write(() => {
    const setToDelete = realm.objectForPrimaryKey(
      WORKOUT_SET_SCHEMA_NAME,
      setId
    );
    realm.delete(setToDelete);
    // After deleting set, check if we need to delete the whole exercise
    const exerciseId = getExerciseSchemaIdFromSet(setId);
    const exercise = realm.objectForPrimaryKey(
      WORKOUT_EXERCISE_SCHEMA_NAME,
      exerciseId
    );
    if (exercise.sets.length === 0) {
      deleteWorkoutExercise(exercise);
    }
  });
};

export const getLastSetByType = (type: ?string) =>
  realm
    .objects(WORKOUT_SET_SCHEMA_NAME)
    .filtered('type = $0', type)
    .sorted([['date', true], ['id', true]]);

export const getSetsThisWeek = (category: ExerciseCategoryType) => {
  const [start, end] = getFirstAndLastWeekday(getToday());
  return realm
    .objects(WORKOUT_SET_SCHEMA_NAME)
    .filtered(
      `category = $0 AND date >= $1 AND date <= $2`,
      category,
      start,
      end
    );
};
