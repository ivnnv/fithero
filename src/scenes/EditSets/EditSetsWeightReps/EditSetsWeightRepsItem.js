/* @flow */

import * as React from 'react';
import { StyleSheet } from 'react-native';
import { Text, useTheme } from 'react-native-paper';

import i18n from '../../../utils/i18n';
import type { WorkoutSetWeightRepsType } from '../../../database/types';
import type { ThemeType } from '../../../utils/theme/withTheme';
import { toLb, toTwoDecimals } from '../../../utils/metrics';
import type { DefaultUnitSystemType } from '../../../redux/modules/settings';

type Props = {|
  set: WorkoutSetWeightRepsType,
  unit: DefaultUnitSystemType,
|};

const EditSetsWeightRepsItem = (props: Props) => {
  const { set, unit } = props;

  const theme: ThemeType = useTheme();

  return (
    <>
      <Text style={[styles.text, styles.weight]}>
        {unit === 'metric'
          ? toTwoDecimals(set.weight)
          : toTwoDecimals(toLb(set.weight))}{' '}
        <Text style={[styles.unit, { color: theme.colors.secondaryText }]}>
          {unit === 'metric'
            ? i18n.t('kg.unit', { count: set.weight })
            : i18n.t('lb')}{' '}
        </Text>
      </Text>
      <Text style={[styles.text, styles.reps]}>
        {set.reps}{' '}
        <Text style={[styles.unit, { color: theme.colors.secondaryText }]}>
          {i18n.t('reps.unit', { count: set.reps })}{' '}
        </Text>
      </Text>
    </>
  );
};

const styles = StyleSheet.create({
  text: {
    fontSize: 18,
  },
  weight: {
    flex: 0.5,
    textAlign: 'right',
  },
  reps: {
    flex: 0.5,
    paddingLeft: 16,
    textAlign: 'right',
  },
  unit: {
    fontSize: 14,
  },
});

export default EditSetsWeightRepsItem;
