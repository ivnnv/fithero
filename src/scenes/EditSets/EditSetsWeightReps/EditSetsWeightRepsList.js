/* @flow */

import React, { useCallback, memo } from 'react';

import type {
  WorkoutExerciseWeightRepsType,
  WorkoutSetSchemaType,
} from '../../../database/types';
import EditSetItem from '../components/EditSetItem';
import useMaxSetHook from '../../../hooks/useMaxSetHook';
import {
  getMaxRepByType,
  getMaxWeightByType,
} from '../../../database/services/WorkoutSetService';
import { REALM_DEFAULT_DEBOUNCE_VALUE } from '../../../database/constants';
import type { DefaultUnitSystemType } from '../../../redux/modules/settings';
import { useSelector } from 'react-redux';
import { getWeightUnit } from '../../../utils/metrics';
import EditSetsList from '../components/EditSetsList';
import EditSetsWeightRepsItem from './EditSetsWeightRepsItem';
import type { ThemeType } from '../../../utils/theme/withTheme';
import { useTheme } from 'react-native-paper';

type Props = {|
  exercise: ?WorkoutExerciseWeightRepsType,
  selectedId: string,
  onPressItem: (setId: string) => void,
  type: string,
|};

const EditSetsWeightRepsList = (props: Props) => {
  const { exercise, onPressItem, selectedId, type } = props;
  const data = exercise ? exercise.sets : [];

  const { colors }: ThemeType = useTheme();
  const defaultUnitSystem: DefaultUnitSystemType = useSelector(
    state => state.settings.defaultUnitSystem
  );

  const unit = getWeightUnit(exercise, defaultUnitSystem);

  const maxSet: ?WorkoutSetSchemaType = useMaxSetHook(
    type,
    getMaxWeightByType,
    REALM_DEFAULT_DEBOUNCE_VALUE
  );
  const maxRep: ?WorkoutSetSchemaType = useMaxSetHook(
    type,
    getMaxRepByType,
    REALM_DEFAULT_DEBOUNCE_VALUE
  );

  const maxSetId = maxSet ? maxSet.id : null;
  const maxRepId = maxRep ? maxRep.id : null;

  const renderItem = useCallback(
    ({ item, index }) => {
      const trophyColor =
        maxSetId === item.id
          ? colors.trophy
          : maxRepId === item.id
          ? colors.trophyReps
          : null;
      return (
        <EditSetItem
          set={item}
          index={index + 1}
          isSelected={selectedId === item.id}
          trophyColor={trophyColor}
          onPressItem={onPressItem}
        >
          <EditSetsWeightRepsItem set={item} unit={unit} />
        </EditSetItem>
      );
    },
    [
      colors.trophy,
      colors.trophyReps,
      maxRepId,
      maxSetId,
      onPressItem,
      selectedId,
      unit,
    ]
  );

  return <EditSetsList data={data} renderItem={renderItem} />;
};

export default memo<Props>(EditSetsWeightRepsList, (prevProps, nextProps) => {
  if (
    prevProps.selectedId !== nextProps.selectedId ||
    prevProps.onPressItem !== nextProps.onPressItem ||
    prevProps.type !== nextProps.type
  ) {
    return false;
  }

  return (
    // Doing it like this because of Realm
    JSON.stringify(prevProps.exercise) === JSON.stringify(nextProps.exercise)
  );
});
