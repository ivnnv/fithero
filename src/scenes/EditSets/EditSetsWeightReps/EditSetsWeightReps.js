/* @flow */

import React, { useCallback, useEffect, useRef, useState } from 'react';
import { Keyboard, StyleSheet, View } from 'react-native';

import type { WorkoutExerciseWeightRepsType } from '../../../database/types';
import EditSetsInputControls from '../components/EditSetsInputControls';
import i18n from '../../../utils/i18n';
import EditSetActionButtons from '../components/EditSetActionButtons';
import { deleteSet } from '../../../database/services/WorkoutSetService';
import {
  getWeight,
  getWeightUnit,
  toKg,
  toTwoDecimals,
} from '../../../utils/metrics';
import type { DefaultUnitSystemType } from '../../../redux/modules/settings';
import Card from '../../../components/Card';
import { useSelector } from 'react-redux';
import { addSetFromInput, getLastSet, getLastWeight } from '../utils';
import usePrevious from '../../../hooks/usePrevious';
import useSelectedId from '../hooks/useSelectedId';
import EditSetsWeightRepsList from './EditSetsWeightRepsList';

type Props = {
  day: string,
  exerciseKey: string,
  exercise: ?WorkoutExerciseWeightRepsType,
  selectedPage?: number,
};

const EditSetsWeightReps = (props: Props) => {
  const { day, exercise, exerciseKey, selectedPage } = props;
  const [selectedId, setSelectedId] = useSelectedId(selectedPage);
  const defaultUnitSystem: DefaultUnitSystemType = useSelector(
    state => state.settings.defaultUnitSystem
  );
  const previousDefaultUnitSystem = usePrevious<DefaultUnitSystemType>(
    defaultUnitSystem
  );
  const isAddingExercise = useRef(false);
  const unit = getWeightUnit(exercise, defaultUnitSystem);
  const lastSet = getLastSet(exercise, exerciseKey);
  const lastWeight = getLastWeight(exercise, lastSet, unit);
  const lastReps = lastSet ? lastSet.reps.toString() : '8';

  const [weight, setWeight] = useState(lastWeight);
  const [reps, setReps] = useState(lastReps);

  const handleIncDec = useCallback(
    (value, setValue, property, increment: number) => {
      const currentValue = value >= '0' ? value : '0';
      const parsedValue =
        property === 'weight'
          ? parseFloat(currentValue)
          : parseInt(currentValue, 10);
      const newValue = (parsedValue + increment).toString();
      setValue(newValue > '0' ? newValue : '0');
    },
    []
  );

  const handleWeightInc = useCallback(() => {
    handleIncDec(weight, setWeight, 'weight', +1);
  }, [handleIncDec, weight]);

  const handleWeightDec = useCallback(() => {
    handleIncDec(weight, setWeight, 'weight', -1);
  }, [handleIncDec, weight]);

  const handleRepsInc = useCallback(() => {
    handleIncDec(reps, setReps, 'reps', +1);
  }, [handleIncDec, reps]);

  const handleRepsDec = useCallback(() => {
    handleIncDec(reps, setReps, 'reps', -1);
  }, [handleIncDec, reps]);

  const onChangeWeightInput = useCallback((value: string) => {
    const parsedValue = value.replace(',', '.');
    // TODO handle comma correctly depending on the locale (save it using dot but showing it using comma)
    if (value === '-' || !isNaN(parsedValue)) {
      setWeight(parsedValue);
    }
  }, []);

  const onChangeRepsInput = useCallback((value: string) => {
    setReps(value === '' || parseInt(value, 10) >= 0 ? value : '0');
  }, []);

  const onPressItem = useCallback(
    (setId: string) => {
      if (selectedId === setId) {
        setSelectedId('');
        return;
      }
      if (exercise) {
        const { sets } = exercise;
        const set = sets.find(s => s.id === setId);
        if (set) {
          setSelectedId(setId);
          setWeight(
            toTwoDecimals(getWeight(set.weight, exercise, unit)).toString()
          );
          setReps(set.reps.toString());
        }
      }
    },
    [exercise, selectedId, setSelectedId, unit]
  );

  const onAddSet = useCallback(() => {
    // If the user presses very fast it can try to create a duplicated primary key
    if (isAddingExercise.current) {
      return;
    }
    isAddingExercise.current = true;

    Keyboard.dismiss();

    let newWeight = 0;
    if (weight !== '' && !isNaN(weight)) {
      newWeight =
        unit === 'metric' ? parseFloat(weight) : toKg(parseFloat(weight));
    }

    const newReps = reps ? parseInt(reps, 10) : 0;

    addSetFromInput({
      day,
      exerciseKey,
      exercise,
      category: 'weight_reps',
      partialSet: {
        weight: newWeight,
        reps: newReps,
      },
      weight_unit: unit,
      updateId: selectedId,
    });

    if (selectedId) {
      setSelectedId('');
    }
  }, [
    day,
    exercise,
    exerciseKey,
    reps,
    selectedId,
    setSelectedId,
    unit,
    weight,
  ]);

  const onDeleteSet = useCallback(() => {
    Keyboard.dismiss();

    deleteSet(selectedId);
    setSelectedId('');
  }, [selectedId, setSelectedId]);

  useEffect(() => {
    if (previousDefaultUnitSystem !== defaultUnitSystem && !exercise) {
      const unit = getWeightUnit(exercise, defaultUnitSystem);
      const lastSet = getLastSet(exercise, exerciseKey);
      const lastWeight = getLastWeight(exercise, lastSet, unit);
      setWeight(lastWeight);
    }
  }, [defaultUnitSystem, exercise, exerciseKey, previousDefaultUnitSystem]);

  useEffect(() => {
    if (isAddingExercise.current) {
      isAddingExercise.current = false;
    }
  }, [exercise]);

  return (
    <View style={styles.container}>
      <Card style={styles.card}>
        <View style={styles.cardContent}>
          <EditSetsInputControls
            value={weight}
            label={i18n.t('weight_label', {
              w:
                unit === 'metric'
                  ? i18n.t('kg.unit', { count: 10 })
                  : i18n.t('lb'),
            })}
            onChangeText={onChangeWeightInput}
            controls={[
              { icon: 'minus', action: handleWeightDec },
              { icon: 'plus', action: handleWeightInc },
            ]}
            keyboardType="numeric"
            containerStyle={[styles.weightContainer, styles.weightSeparation]}
            labelStyle={styles.weightSeparation}
            testID="weightInput"
          />
          <EditSetsInputControls
            value={reps}
            label={i18n.t('reps.title')}
            onChangeText={onChangeRepsInput}
            controls={[
              { icon: 'minus', action: handleRepsDec },
              { icon: 'plus', action: handleRepsInc },
            ]}
            keyboardType="number-pad"
            containerStyle={[styles.repsContainer, styles.repsSeparation]}
            labelStyle={styles.repsSeparation}
            testID="repsInput"
          />
        </View>
        <EditSetActionButtons
          isUpdate={!!selectedId}
          onAddSet={onAddSet}
          onDeleteSet={onDeleteSet}
        />
      </Card>
      <EditSetsWeightRepsList
        exercise={exercise}
        onPressItem={onPressItem}
        selectedId={selectedId}
        type={exerciseKey}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  card: {
    marginTop: 8,
    marginHorizontal: 16,
  },
  cardContent: {
    flexDirection: 'row',
    justifyContent: 'center',
    paddingTop: 8,
    paddingHorizontal: 8,
  },
  weightContainer: {
    flex: 1,
  },
  repsContainer: {
    flex: 0.9,
  },
  weightSeparation: {
    paddingRight: 8,
  },
  repsSeparation: {
    paddingLeft: 8,
  },
});

export default EditSetsWeightReps;
