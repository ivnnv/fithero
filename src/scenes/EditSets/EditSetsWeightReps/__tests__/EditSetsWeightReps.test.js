/* @flow */

import React from 'react';
import { Keyboard } from 'react-native';
import { Provider } from 'react-redux';
import { fireEvent, render } from 'react-native-testing-library';
import { useRoute } from '@react-navigation/native';

import EditSetsWeightReps from '../index';
import { toDate } from '../../../../utils/date';
import {
  addSet,
  deleteSet,
  getLastSetByType,
  getMaxSetByType,
  updateSet,
} from '../../../../database/services/WorkoutSetService';
import { MockRealmArray } from '../../../../database/services/__tests__/helpers/databaseMocks';
import { createStore } from 'redux';
import { addExercise } from '../../../../database/services/WorkoutExerciseService';

jest.mock('react-native/Libraries/Components/Keyboard/Keyboard');
jest.mock('../../../../database/services/WorkoutSetService', () => ({
  addSet: jest.fn(),
  deleteSet: jest.fn(),
  getLastSetByType: jest.fn(() => []),
  updateSet: jest.fn(),
  getMaxSetByType: jest.fn(),
  getMaxRepByType: jest.fn(),
}));
jest.mock('../../../../database/services/WorkoutExerciseService', () => ({
  addExercise: jest.fn(),
}));
jest.mock('../../../../hooks/useKeyboard');
jest.mock('../../../../hooks/useMaxSetHook');

const date = toDate('2018-05-01T00:00:00.000Z');

describe('EditSetsWeightReps', () => {
  const day = '2018-05-01T00:00:00.000Z';
  const exerciseKey = 'bench-press';
  const exercise = {
    id: '20180501_bench-press',
    sets: [
      {
        id: '20180501_bench-press_001',
        reps: 8,
        weight: 80,
        date,
        type: 'bench-press',
      },
      {
        id: '20180501_bench-press_002',
        reps: 6,
        weight: 90,
        date,
        type: 'bench-press',
      },
    ],
    date,
    type: 'bench-press',
    sort: 0,
    weight_unit: 'metric',
    isValid: jest.fn(() => true),
  };

  const defaultWeight = 20;
  const defaultReps = 8;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  const _render = (props, defaultUnitSystem = 'metric') => {
    useRoute.mockReturnValue({
      params: {
        day,
        exerciseKey,
      },
    });
    return render(
      <Provider
        store={createStore(() => ({
          settings: { defaultUnitSystem },
        }))}
      >
        <EditSetsWeightReps
          selectedPage={0}
          exercise={null}
          // $FlowIgnore
          {...props}
        />
      </Provider>
    );
  };

  describe('EditSetsInputControls', () => {
    it('has correct default values if there is no exercise', async () => {
      const { getByTestId } = _render();

      const weightInput = getByTestId('weightInput');
      const repsInput = getByTestId('repsInput');
      const weightLabel = getByTestId('weightInputLabel');
      const repsLabel = getByTestId('repsInputLabel');

      expect(weightInput.props.value).toEqual(defaultWeight.toString());
      expect(repsInput.props.value).toEqual(defaultReps.toString());
      expect(weightLabel.props.children).toEqual('Weight (kgs)');
      expect(repsLabel.props.children).toEqual('Reps');
    });

    it('has the values of last set if we pass an exercise', () => {
      const { getByTestId } = _render({
        exercise: {
          weight_unit: 'metric',
          sort: 1,
          isValid: jest.fn(),
          ...exercise,
        },
      });

      const weightInput = getByTestId('weightInput');
      const repsInput = getByTestId('repsInput');

      expect(weightInput.props.value).toEqual(
        exercise.sets[1].weight.toString()
      );
      expect(repsInput.props.value).toEqual(exercise.sets[1].reps.toString());
    });

    it('has values of last set (from another day) if no exercise', () => {
      const mockLastSet = {
        id: '2018-05-01T00:00:00.000Z_bench-press_002',
        reps: 6,
        weight: 90,
        date: toDate('2018-05-01T00:00:00.000Z'),
        type: 'bench-press',
      };

      // $FlowFixMe Flow does not now this is a mock
      getLastSetByType.mockImplementation(() => [mockLastSet]);

      const { getByTestId } = _render();

      const weightInput = getByTestId('weightInput');
      const repsInput = getByTestId('repsInput');

      expect(weightInput.props.value).toEqual(mockLastSet.weight.toString());
      expect(repsInput.props.value).toEqual(mockLastSet.reps.toString());
    });

    it('changes input(s) state using the TextInput', () => {
      const { getByTestId } = _render();

      const weightInput = getByTestId('weightInput');
      const repsInput = getByTestId('repsInput');

      fireEvent.changeText(weightInput, '50');
      fireEvent.changeText(repsInput, '5');

      expect(weightInput.props.value).toEqual('50');
      expect(repsInput.props.value).toEqual('5');
    });

    it.skip('handles empty TextInput', () => {
      // TODO write test
    });

    it('uses -2, -1, +1, +2 buttons for reps', () => {
      // $FlowFixMe Flow does not now this is a mock
      getLastSetByType.mockImplementation(() => []);

      const { getByTestId } = _render();

      const repsInput = getByTestId('repsInput');
      const repsInputControlLeft = getByTestId('repsInputControlLeft');
      const repsInputControlRight = getByTestId('repsInputControlRight');

      fireEvent.press(repsInputControlLeft);
      expect(repsInput.props.value).toEqual((defaultReps - 1).toString());
      fireEvent.press(repsInputControlRight);
      fireEvent.press(repsInputControlRight);
      expect(repsInput.props.value).toEqual(
        (defaultReps - 1 + 1 + 1).toString()
      );
    });

    it('uses -1.0, -0.5, +0.5, +1.0 buttons for weight', () => {
      // $FlowFixMe Flow does not now this is a mock
      getLastSetByType.mockImplementation(() => []);

      const { getByTestId } = _render();

      const weightInput = getByTestId('weightInput');
      const weightInputControlLeft = getByTestId('weightInputControlLeft');
      const weightInputControlRight = getByTestId('weightInputControlRight');

      fireEvent.press(weightInputControlLeft);
      expect(weightInput.props.value).toEqual((defaultWeight - 1).toString());
      fireEvent.press(weightInputControlRight);
      fireEvent.press(weightInputControlRight);
      expect(weightInput.props.value).toEqual(
        (defaultWeight - 1 + 1 + 1).toString()
      );
    });
  });

  describe('EditSetActionButtons and back button', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    it('switches between Add and Update text if a set is selected', async () => {
      const { getByTestId } = _render({ exercise });
      expect(getByTestId('addSetButton').props.children).toBe('Add');
      fireEvent.press(getByTestId('editSetItem-1'));
      expect(getByTestId('addSetButton').props.children).toBe('Update');
    });

    it('switches Delete button to enabled/disabled depending on set selection', () => {
      const { getByTestId } = _render({ exercise });
      // Nothing selected
      expect(getByTestId('deleteSetButton').props.disabled).toBe(true);
      // Something selected means delete button is enabled
      fireEvent.press(getByTestId('editSetItem-1'));
      expect(getByTestId('deleteSetButton').props.disabled).toBe(false);
    });

    it('deletes a set and dismiss the keyboard', () => {
      const { getByTestId } = _render();

      expect(Keyboard.dismiss).not.toHaveBeenCalled();

      fireEvent.press(getByTestId('deleteSetButton'));

      expect(Keyboard.dismiss).toHaveBeenCalled();
      expect(deleteSet).toHaveBeenCalledTimes(1);
    });
  });

  describe('Weight units', () => {
    // $FlowFixMe
    getMaxSetByType.mockImplementation(
      () => new MockRealmArray({ ...exercise.sets[0] })
    );

    it('renders kgs or lbs', () => {
      const { toJSON: metricJSON } = _render({ exercise });
      const { toJSON: imperialJSON } = _render({
        exercise: {
          ...exercise,
          weight_unit: 'imperial',
        },
      });

      // $FlowFixMe
      expect(metricJSON()).toMatchDiffSnapshot(imperialJSON(), {
        contextLines: 0,
        stablePatchmarks: true,
      });
    });
  });

  describe('onAddSet', () => {
    it('dismissed the keyboard when adding a set', () => {
      const { getByTestId } = _render();

      expect(Keyboard.dismiss).not.toHaveBeenCalled();

      fireEvent.press(getByTestId('addSetButton'));

      expect(Keyboard.dismiss).toHaveBeenCalled();
    });

    const _addSet = getByTestId => {
      const weightInput = getByTestId('weightInput');
      const repsInput = getByTestId('repsInput');

      fireEvent.changeText(weightInput, '100');
      fireEvent.changeText(repsInput, '6');

      fireEvent.press(getByTestId('addSetButton'));
    };

    it('creates a new exercise with a new set if there is no exercise', () => {
      const { getByTestId } = _render();
      _addSet(getByTestId);
      fireEvent.press(getByTestId('addSetButton'));

      expect(addExercise).toHaveBeenCalledWith({
        id: '20180501_bench-press',
        date,
        category: 'weight_reps',
        sets: [
          {
            id: '20180501_bench-press_001',
            date,
            weight: 100,
            reps: 6,
            category: 'weight_reps',
            type: 'bench-press',
          },
        ],
        weight_unit: 'metric',
        type: 'bench-press',
      });
    });

    it('adds a set to an existing exercise', () => {
      const { getByTestId } = _render({ exercise });
      _addSet(getByTestId);
      fireEvent.press(getByTestId('addSetButton'));

      expect(addExercise).not.toHaveBeenCalled();
      expect(addSet).toHaveBeenCalledWith({
        id: '20180501_bench-press_003',
        date,
        weight: 100,
        reps: 6,
        category: 'weight_reps',
        type: 'bench-press',
      });
    });

    it('updates an existing set', async () => {
      const { getByTestId } = _render({ exercise });
      fireEvent.press(getByTestId('editSetItem-1'));
      _addSet(getByTestId);

      fireEvent.press(getByTestId('addSetButton'));

      expect(addExercise).not.toHaveBeenCalled();
      expect(addSet).not.toHaveBeenCalled();
      expect(updateSet).toHaveBeenCalledWith({
        id: exercise.sets[0].id,
        date,
        weight: 100,
        reps: 6,
        category: 'weight_reps',
        type: 'bench-press',
      });
    });
  });
});
