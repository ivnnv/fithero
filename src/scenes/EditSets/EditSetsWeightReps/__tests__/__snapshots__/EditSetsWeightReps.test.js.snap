// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSetsWeightReps Weight units renders kgs or lbs 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ --- --- @@
-             Weight (kgs)
+             Weight (lb)
@@ --- --- @@
-               value="90"
+               value="198.42"
@@ --- --- @@
-             80
+             176.37
@@ --- --- @@
-               kgs
+               lb
@@ --- --- @@
-             90
+             198.42
@@ --- --- @@
-               kgs
+               lb
`;
